ntcore/src/main/native/include/networktables/ProtobufTopic.h

```diff
@@ -0,0 +1,415 @@
+ */
+template <wpi::ProtobufSerializable T>
+class ProtobufEntry final : public ProtobufSubscriber<T>,
+                            public ProtobufPublisher<T> {
+ public:
+  using SubscriberType = ProtobufSubscriber<T>;
+  using PublisherType = ProtobufPublisher<T>;
+  using TopicType = ProtobufTopic<T>;
+  using ValueType = T;
+  using ParamType = const T&;
+
+  using TimestampedValueType = TimestampedObject<T>;
+
+  ProtobufEntry() = default;
+
+  /**
+   * Construct from an entry handle; recommended to use
+   * ProtobufTopic::GetEntry() instead.
+   *
+   * @param handle Native handle
+   * @param defaultValue Default value
+   */
+  ProtobufEntry(NT_Entry handle, T defaultValue)
+      : ProtobufSubscriber<T>{handle, std::move(defaultValue)},
+        ProtobufPublisher<T>{handle} {}
+
+  /**
+   * Determines if the native handle is valid.
+   *
+   * @return True if the native handle is valid, false otherwise.
+   */
+  explicit operator bool() const { return this->m_subHandle != 0; }
+
+  /**
+   * Gets the native handle for the entry.
+   *
+   * @return Native handle
+   */
+  NT_Entry GetHandle() const { return this->m_subHandle; }
+
+  /**
+   * Get the corresponding topic.
+   *
+   * @return Topic
+   */
+  TopicType GetTopic() const {
+    return ProtobufTopic<T>{::nt::GetTopicFromHandle(this->m_subHandle)};
+  }
+
+  /**
+   * Stops publishing the entry if it's published.
+   */
+  void Unpublish() { ::nt::Unpublish(this->m_pubHandle); }
+};
+
+/**
+ * NetworkTables protobuf-encoded value topic.
+ */
+template <wpi::ProtobufSerializable T>
+class ProtobufTopic final : public Topic {
+ public:
+  using SubscriberType = ProtobufSubscriber<T>;
+  using PublisherType = ProtobufPublisher<T>;
+  using EntryType = ProtobufEntry<T>;
+  using ValueType = T;
+  using ParamType = const T&;
+  using TimestampedValueType = TimestampedObject<T>;
+
+  ProtobufTopic() = default;
+
+  /**
+   * Construct from a topic handle; recommended to use
+   * NetworkTableInstance::GetProtobufTopic() instead.
+   *
+   * @param handle Native handle
+   */
+  explicit ProtobufTopic(NT_Topic handle) : Topic{handle} {}
+
+  /**
+   * Construct from a generic topic.
+   *
+   * @param topic Topic
+   */
+  explicit ProtobufTopic(Topic topic) : Topic{topic} {}
+
+  /**
+   * Create a new subscriber to the topic.
+   *
+   * <p>The subscriber is only active as long as the returned object
+   * is not destroyed.
+   *
+   * @note Subscribers that do not match the published data type do not return
+   *     any values. To determine if the data type matches, use the appropriate
+   *     Topic functions.
+   *
+   * @param defaultValue default value used when a default is not provided to a
+   *        getter function
+   * @param options subscribe options
+   * @return subscriber
+   */
+  [[nodiscard]] SubscriberType Subscribe(
+      T defaultValue, const PubSubOptions& options = kDefaultPubSubOptions) {
```

Comment:
Why is this `T` (`ValueType`) and not `const T&` (`ParamType`)? (Also in `ProtobufTopic::GetEntry`)

Reply:
When constructing objects (and this delegates to a constructor), the recommended approach is to pass T and std::move it into the member, as this can avoid an unnecessary copy (the compiler can see the move and construct in place).  See https://clang.llvm.org/extra/clang-tidy/checks/modernize/pass-by-value.html